outline/outline
# Feature: Add iframe Insertion Capability to the Editor

### Description

This PR introduces a new feature that allows users to insert iframes into the editor. Based on the discussions in #5137, various adjustments have been made to enhance the implementation. The iframe is now inserted directly into the document's source code as an `<iframe>` HTML tag. This ensures that the content is rendered correctly even when the document is accessed via API for embedding in other pages.


### Changes

- Added a new toolbar button for inserting iframes.
- Updated the editor's parsing and rendering logic to handle direct iframe HTML tags.
- Added a ﻿`urlTransformer` function configuration to the `﻿EmbedDescriptor` to support transforming user-entered URLs into predefined text content, such as wrapping them with HTML tags. This approach avoids matching all external content by domain and allows for future enhancements to embed more diverse content types.
- Ensured iframes remain functional and correctly displayed when the document content is fetched and rendered in other applications via API.

### Screenshots

<img width="1287" alt="Bevis_2024-05-26_at_23 35 33@2x" src="https://github.com/outline/outline/assets/2999386/65169e33-9386-4793-9d52-1d84230f7c09">

### How to Test

1. Open the editor.
2. Press "/"
3. Select the new iframe button in the menu.
4. Enter a valid iframe URL and any additional attributes in the dialog.
5. Insert the iframe and verify it is correctly embedded and displayed.
6. Save the document and access its content via the API to ensure the iframe is properly rendered in other applications.

### Others

I believe that solely relying on URL matching presents several issues. For scenarios where external API requests are needed to render document content, it adds some pressure. It would be best to consider wrapping all external content with iframe HTML tags in future updates, reflected in the document's original Markdown content.

---

Feel free to adjust details according to the specific implementation and requirements of the project!
